### PR TITLE
Fixed #20625 -- Custom querysets can be chained multiple times.

### PR DESCRIPTION
**Pull request by coady**
https://code.djangoproject.com/ticket/20625#comment:39

### CODE REVIEW

**Comment by timgraham** [comment]
I don't know that we need to fix this on 1.8, but I guess if the other test failure can be resolved without too much difficult, it would be okay. We should have a separate Trac ticket though.

File: django/db/models/query.py [hunk 1]
```diff
@@ -937,10 +937,10 @@ def _batched_insert(self, objs, fields, batch_size):
                                              using=self.db)
 
     def _clone(self, klass=None, setup=False, **kwargs):
+        base_queryset_class = getattr(self, '_base_queryset_class', self.__class__)
         if klass is None:
             klass = self.__class__
-        elif not issubclass(self.__class__, klass):
-            base_queryset_class = getattr(self, '_base_queryset_class', self.__class__)
+        elif not (issubclass(base_queryset_class, klass) or issubclass(klass, base_queryset_class)):
```
Comment (loic) [inline django/db/models/query.py]: Why not just `elif not issubclass(base_queryset_class, klass):`?

Comment (acoady) [inline django/db/models/query.py]: The goal state is `issubclass(klass, base_queryset_class)`.  The additional check is to handle the edge case of ValueListQuerySet being *upcasted* to ValuesQuerySet.

Comment (timgraham) [inline django/db/models/query.py]: Could you add a test for that case?

Comment (coady) [inline django/db/models/query.py]: Done.

**Comment by timgraham** [comment]
merged in f33b3ebd5302fe5821dd796156b0be66c5aebec9, thanks!
